### PR TITLE
Add Expect.within

### PR DESCRIPTION
**Pull request by mgold**
Add `Expect.within` as discussed in #41. However, I'd like to merge #85 first so this can be implemented in the new style.

### CODE REVIEW

**Comment by mgold** [comment]
This is ready for review now.

File: src/Expect.elm [hunk 1]
```diff
@@ -234,6 +235,39 @@ atLeast =
     compareWith "Expect.atLeast" (>=)
 
 
+{-| Passes if the second and third arguments are equal within a tolerance
+specified by the first argument. This is indended to avoid failing because of
+minor inaccuracies introduced by floating point arithmetic.
+
+    -- Fails because 0.1 + 0.2 == 0.30000000000000004 (yes, seriously)
```
Comment (drathier) [inline src/Expect.elm]: `(yes, seriously)` -> `(0.1 is irrational in base 2)`

**Comment by rtfeldman** [comment]
@drathier sorry - can you clarify what you mean about relative error? I'm not sure what part of the discussion in #41 you're referring to.

File: src/Expect.elm [hunk 2]
```diff
@@ -234,6 +235,39 @@ atLeast =
...
+    │ Expect.within 0.0001
+    ╵
+    3.141592653589793
+
+    -}
+-}
+within : Float -> Float -> Float -> Expectation
+within tolerance =
+    compareWith ("Expect.within " ++ toString tolerance)
+        (\a b -> abs (a - b) < tolerance)
```
Comment (rtfeldman) [inline src/Expect.elm]: I'd expect this to be `<=` instead of `<`

(That is, I'd expect it to be true that `0.5` is within `0.1` of `0.4`)

**Comment by drathier** [comment]
@rtfeldman If you're adding two floats, what tolerance should you have? If the floats are huge, e.g. 10^20, compared to really small, 10^-20, you would want different tolerances. 10^20±10^4 is probably fine, but 10^-10±10^4 absolutely isn't. Absolute is all you need for constants and small ranges, but when you cover larger ranges, you need a relative tolerance, like ±1%.

**Comment by rtfeldman** [comment]
I don't think we should add this to the API until we figure out what to do about ☝️

**Comment by drathier** [comment]
I took a shot at it in #93

File: tests/Helpers.elm [hunk 3]
```diff
@@ -23,6 +23,16 @@ expectToFail =
     expectFailureHelper (always Nothing)
 
 
+didFail : Expectation -> Bool
```
Comment (drathier) [inline tests/Helpers.elm]: I'd like to negate this function. How about `succeeded`?

Comment (mgold) [inline tests/Helpers.elm]: f2d46d6

**Comment by mgold** [comment]
@drathier Running the tests locally, I have one test failure. I have to go but please take a look and suggest or PR a fix.

**Comment by mgold** [comment]
@rtfeldman Any idea why the error message printout isn't telling us which test failed? That's rather useful information...

**Comment by drathier** [comment]
I am doing some maths to figure out how we should handle the near-zero case. I'm not pleased with the current implementation. Not sure if it'll be done before the holidays.

**Comment by rtfeldman** [comment]
No rush! Appreciate the attention to detail.

**Comment by rtfeldman** [comment]
Should we close this in favor of #102?

**Comment by drathier** [comment]
@rtfeldman Yes :)

**Comment by rtfeldman** [comment]
Closed in favor of https://github.com/elm-community/elm-test/pull/102

**Comment by drathier** [comment]
@rtfeldman #102 merged into this branch, so please merge this one too :) I forked this PR when I created that other PR.

**Comment by rtfeldman** [comment]
@drathier If I'm understanding, that should mean since #102 is merged, these commits are now merged already, yeah?

**Comment by drathier** [comment]
Yeah, #102 merged into this PR branch, but not into master, so now we want to merge this into master. Look at the commit history for this PR: "@rtfeldman	Merge pull request #102 from drathier/master  …	fade2ce". 

Since I didn't have access to push to this branch before, I forked it. Then I submitted a PR to merge it back where I forked it from, i.e. this PR. I should have asked to merge it into master instead, but I didn't. Then just after the merge, you deleted this branch before Travis could fetch it, so all tests failed. Sorry for the confusing merges :)

Something like this should do it:
1. restore branch
2. reopen PR
3. run travis tests just to be sure (trigger them manually at https://travis-ci.org/elm-community/elm-test/builds/195381778)
4. merge PR

I could do it myself, but I don't want to merge my own PR.

**Comment by mgold** [comment]
Waiting on CI, even though tests passed locally.

**Comment by mgold** [comment]
CI seems to be not starting the build on macOS. Merging anyway...

**Comment by drathier** [comment]
@mgold yes, that review was very old. Thanks for breaking out the tests into multiple files.

**Pull request merged**